### ISSUE TITLE
HFlow changes layout behaviour of TextField

### ISSUE THREAD

**Issue by lhunath** [tevelee/SwiftUI-Flow#21]
consider:
```swift
VStack {
    HFlow {
        TextField("Test", text: .constant(""))
            .border(Color.red)
    }
    TextField("Test", text: .constant(""))
        .border(Color.red)
}
```
We notice that `HFlow` is affecting the default layout behaviour of `TextField`:
<img width="411" alt="image" src="https://github.com/user-attachments/assets/b78ff8c5-5a1a-4bef-b43a-fd9a22f54b4d">
A `TextField` wants to expand horizontally to adopt all available space, but the `HFlow` is disallowing this.
Unless there is some mechanism in place to restore the default behaviour of the `TextField` component (and likely others?), `HFlow` should be adjusted to restore it.

**Comment by tevelee** [comment]
Try setting the alignment in the `VStack` explicitly. 

```swift
VStack(alignment: .leading) {
    HFlow {
        TextField("Test", text: .constant(""))
            .border(Color.red)
    }
    TextField("Test", text: .constant(""))
        .border(Color.red)
}
```

![image](https://github.com/user-attachments/assets/a11e188d-3968-4fe1-9cac-69aa6eb0baaa)

**Comment by lhunath** [comment]
Thanks, I am curious - does this not take away some agency of the individual items in the flow? Some views are naturally content-wrapping, while some views are naturally expanding. It would appear this behaviour overrides these natural preferences and either makes all items wrapping or all expanding, regardless of their individual nature.
One might expect, eg. `Text("Test")` to be naturally content-wrapped, but `TextField("Test", _)` to be horizontally-expanding. As designed, these two items with their natural preferences cannot be honoured within the same flow.
Is there a reason we cannot honour the views' default preferences?

**Comment by tevelee** [comment]
Thank you for the thoughtful question @lhunath! This opens up an interesting exploration of layout behavior.

`HFlow` is a size-constrained layout that prompts its child views for their ideal sizes and arranges them in rows based on that information. It already respects each view’s sizing preferences by proposing constrained sizes, and it’s up to the child views to respond accordingly.

Regarding flexible views like `TextField`, consider a case with multiple of them in the same `Flow`. Should they all fit in one row, or should they wrap into separate rows? This decision often depends on the layout context, and `HFlow` leaves that choice to the user rather than enforcing assumptions about flexibility.

The default design of `HFlow` prioritizes common use cases like arranging tags, texts, or other _small_ items in rows. For flexible views, there’s ambiguity: where do they fit best within a constrained flow layout? While respecting flexibility might seem desirable in certain cases, it could lead to unintuitive layouts, especially when multiple flexible views are involved.

(Additionally, the library provides an option to evenly distribute items using the Knuth-Plass line-breaking algorithm. Dynamically handling flexible subview constraints in this mode involves a lot of ambiguity, potential conflicts and significant complexity. Layout priorities can possibly reduce some of those ambiguities.)

That said, I’d love to see an example where preserving the flexible nature of views (in the default case) would lead to a better outcome. How would you handle cases with (potentially multiple) flexible views?

Looking forward to your thoughts! This could help shape the future direction of the library.

**Comment by lhunath** [comment]
Appreciate expanding on this, @tevelee. What you say makes sense.

Given that `HFlow` sizes the items based on their own sizing preferences within a constrained context, might I inquire as to whether enough information is passed to the child in this case for it to elect to size itself to the full available (remaining) width in the row? How might I implement such a custom view?

Or is it the case that within this constrained sizing environment, the child has no information on the parent's width, nor does it have an option to size itself non-content-wrapping?

Certainly one option on the table would be a custom modifier which can be selectively applied to children in order to indicate to the `HFlow` that this particular child wishes to be measured differently (though I am not sure about the complexity of implementing such).

**Comment by tevelee** [comment]
I've been thinking about this issue in the past 2 days, and right now I think what you're proposing is possible but very complex to implement. I've been trying to prototype the idea of arranging the flexible items in either compact, natural, or expanded modes. In compact mode (which is essentially current behavior) items are squished. Natural is what you are proposing, adding the flexible item and letting it span however possible. Expanded makes room for flexible items to spread even full rows.

The implementation is pretty complex at this point, but I'm trying to wrap my head around it. The current API has to change to adopt a new mode like this, since this will redefine the justified behavior and the line breaking. (I've already considered some breaking changes for 2.0).

**Comment by lhunath** [comment]
Best of luck, and much appreciation for the consideration. I'm around to feedback into your thought process or think along with some of the concepts should that be beneficial.
Unfortunately I've struggled to find reliable in-depth information from Apple on how the measuring process truly works under the hood, or even a conceptual overview.

**Comment by tevelee** [comment]
I just published version 3.0.0 of the library. Now the default mode of flexibility is `.natural`, the behavior you suggested (with other options to specify). Please try it out if you have a chance

**Comment by lhunath** [comment]
Seems to work beautifully with my initial tests, thanks for the energy!
